types/hapi__joi/index.d.ts

```diff
@@ -152,1245 +226,1907 @@ export interface DomainOptions {
+
+        /**
+         * Sets the presence mode for the schema.
+         */
+        presence(mode: PresenceMode): this;
+
+        /**
+         * Outputs the original untouched value instead of the casted value.
+         */
+        raw(enabled?: boolean): this;
+
+        /**
+         * Marks a key as required which will not allow undefined as value. All keys are optional by default.
+         */
+        required(): this;
+
+        /**
+         * Applies a set of rule options to the current ruleset or last rule added.
+         *
+         * When applying rule options, the last rule (e.g. `min()`) is used unless there is an active ruleset defined (e.g. `$.min().max()`)
+         * in which case the options are applied to all the provided rules.
+         * Once `rule()` is called, the previous rules can no longer be modified and any active ruleset is terminated.
+         *
+         * Rule modifications can only be applied to supported rules.
+         * Most of the `any` methods do not support rule modifications because they are implemented using schema flags (e.g. `required()`) or special
+         * internal implementation (e.g. `valid()`).
+         * In those cases, use the `any.messages()` method to override the error codes for the errors you want to customize.
+         */
+        rule(options: RuleOptions): this;
+
+        /**
+         * Registers a schema to be used by decendents of the current schema in named link references.
+         */
+        shared(ref: Schema): this;
+
+        /**
+         * Sets the options.convert options to false which prevent type casting for the current key and any child keys.
+         */
+        strict(isStrict?: boolean): this;
+
+        /**
+         * Marks a key to be removed from a resulting object or array after validation. Used to sanitize output.
+         * @param [enabled=true] - if true, the value is stripped, otherwise the validated value is retained. Defaults to true.
+         */
+        strip(enabled?: boolean): this;
+
+        /**
+         * Annotates the key
+         */
+        tag(...tags: string[]): this;
+
+        /**
+         * Applies any assigned target alterations to a copy of the schema that were applied via `any.alter()`.
+         */
+        tailor(targets: string | string[]): Schema;
+
+        /**
+         * Annotates the key with an unit name.
+         */
+        unit(name: string): this;
+
+        /**
+         * Adds the provided values into the allowed whitelist and marks them as the only valid values allowed.
+         */
+        valid(...values: any[]): this;
+
+        /**
+         * Validates a value using the schema and options.
+         */
+        validate<T>(value: T, options?: ValidationOptions): ValidationResult<T>;
+
+        /**
+         * Validates a value using the schema and options.
+         */
+        validateAsync<T>(value: T, options?: AsyncValidationOptions): Promise<ValidationResult<T>>;
```

Comment:
This should be:
```
validateAsync<T>(value: T, options?: AsyncValidationOptions): Promise<T>;
```

Reply:
Technically `value` should be `any`, not `T` since we may validate any object type and the schema returns a different one.

Reply:
Right, though you could overwrite it via validateAsync<any> but actually my point was that the result is not a ValidationResult but the value itself.

Reply:
Good point!